library/pkcs7.c

```diff
@@ -664,9 +791,74 @@ static int mbedtls_pkcs7_data_or_hash_verify(mbedtls_pkcs7 *pkcs7,
             continue;
         }
 
-        ret = mbedtls_pk_verify(&pk_cxt, md_alg, hash,
-                                mbedtls_md_get_size(md_info),
-                                signer->sig.p, signer->sig.len);
+        if (signer->signed_attrs_raw.len > 0)
+        {
+            ret = pkcs7_signed_attrs_match(&signer->signed_attrs, hash,
+                                           mbedtls_md_get_size(md_info));
+            if (ret != 0) {
+                mbedtls_free(hash);
+                continue;
+            }
+
+            mbedtls_md_init(&md_ctx);
+            if (mbedtls_md_setup(&md_ctx, md_info, 0) != 0 ||
+                mbedtls_md_starts(&md_ctx) != 0 ||
+                mbedtls_md_update(&md_ctx, &(unsigned char){0x31}, 1) != 0 ||
+                mbedtls_md_update(&md_ctx, signer->signed_attrs_raw.p + 1,
+                                  signer->signed_attrs_raw.len - 1) != 0 ||
+                mbedtls_md_finish(&md_ctx, hash) != 0) {
+                mbedtls_md_free(&md_ctx);
+                mbedtls_free(hash);
+                ret = MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+                continue;
+            }
+            mbedtls_md_free(&md_ctx);
+        }
+
+        ret = MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+
+        /* try trusted certificates */
+        for (pk_cert = pkcs7_find_cert(&signer->serial, cert);
+             pk_cert != NULL;
+             pk_cert = pkcs7_find_cert(&signer->serial, pk_cert->next)) {
+            if (mbedtls_x509_time_is_past(&pk_cert->valid_to) ||
+                mbedtls_x509_time_is_future(&pk_cert->valid_from)) {
+                ret = MBEDTLS_ERR_PKCS7_CERT_DATE_INVALID;
+                continue;
+            }
+            pk_ctx = pk_cert->pk;
+            ret = mbedtls_pk_verify(&pk_ctx, md_alg, hash,
+                                    mbedtls_md_get_size(md_info),
+                                    signer->sig.p, signer->sig.len);
+            if (ret == 0) {
+                break;
+            }
+            ret = MBEDTLS_ERR_PKCS7_VERIFY_FAIL;
+        }
+
+        /* try embedded certs with a chain of trust back to a trusted cert */
+        if (ret != 0) {
+            for (pk_cert = pkcs7_find_cert(&signer->serial,
+                                           &pkcs7->signed_data.certs);
+                 pk_cert != NULL;
+                 pk_cert = pkcs7_find_cert(&signer->serial, pk_cert->next)) {
+                if (mbedtls_x509_crt_verify(pk_cert, cert,
```

Comment:
It seems a non-const `cert` is ONLY required by the call to `mbedtls_x509_crt_verify` here. I'm not sure about this, maybe parameters in `mbedtls_x509_crt_*verify` family:
https://github.com/Mbed-TLS/mbedtls/blob/3c0e4effe4304d3cad09a82126b7939940cd5a2a/library/x509_crt.c#L3110-L3122
should be changed to const at the same time too (which will also change `x509_crt_verify_chain` etc)?